### DISCUSSION TITLE
Tramway Form. Implement callbacks

### DISCUSSION THREAD

**Discussion by kalashnikovisme** [Purple-Magic/tramway#47] [General]
For example, we have something like this. We want to `activate` user after submit

```ruby
class UserForm < Tramway::BaseForm
  def submit(params)
    super.tap do |result|
      object.activate if result == true
    end
  end
end
```

Maybe we should implement callbacks in Tramway Form to be able do something like this?

```ruby
class UserForm < Tramway::BaseForm
  after_submit -> { object.activate }
  after_resist -> { object.deactivate }
end
```

I think it will allow to us do something this too

```ruby
class Users::RejectForm < Tramway::BaseForm
  properties :reject_reason

  after_submit -> { user.reject }
end
```

What do you think?

**Comment by kalashnikovisme** [comment]
@haukot @moshinaan

**Comment by kalashnikovisme** [comment]
One more idea!!

Add option to make callbacks in one transaction with submit!

I mean something like this

```ruby
class Users::RejectForm < Tramway::BaseForm
  properties :reject_reason

  after_submit -> { user.reject! }, single_transaction: true
end
```

in this case, it will run like this

```ruby
ActiveRecord::Base.transaction do
  user.update! params
  user.reject!
end
```

WDYT? @haukot @moshinaan

**Reply by haukot** [reply]
Yes looking good. I saw some gem that runs these callbacks in the same transaction by default

**Reply by moshinaan** [reply]
Looks interesting, could you explain for what cases u need it in this project? I0lhut

**Reply by kalashnikovisme** [reply]
@moshinaan you are asking about callbacks or single_transaction stuff? :slightly_smiling_face:

**Reply by moshinaan** [reply]
About transaction

**Reply by kalashnikovisme** [reply]
For data consistency.

For example,

```ruby
class RejectUserForm < Tramway::BaseForm
  properties :reject_reason

  after_submit -> { user.reject! }, single_transaction: true
end
```

imagine, that `user.reject!` was not processed for some reason. In this case, `reject_reason` will be already committed to database and we will have an object with reject_reason but not in rejected state. With single transaction updating of `reject_reason` would rollbacked.
